y2038: build fails with -D_TIME_BITS=64 for 32-bit targets

`CFLAGS="-D_TIME_BITS=64 -D_FILE_OFFSET_BITS=64 -m32" LDFLAGS=-m32 make -k -j4`

```
src/sfptpd_logging.c: In function ‘sfptpd_log_get_time’:
src/sfptpd_logging.c:1080:63: error: format ‘%ld’ expects argument of type ‘long int’, but argument 5 has type ‘__suseconds64_t’ {aka ‘long long int’} [-
Werror=format=]
 1080 |         rc = snprintf(time->time, sizeof(time->time), "%s.%06ld", temp, now.tv_usec);
      |                                                           ~~~~^         ~~~~~~~~~~~
      |                                                               |            |
      |                                                               long int     __suseconds64_t {aka long long int}
      |                                                           %06lld
src/ptp/ptpd2/dep/msg.c: In function ‘msgDump’:
src/ptp/ptpd2/dep/msg.c:3037:46: error: format ‘%ld’ expects argument of type ‘long int’, but argument 5 has type ‘__suseconds64_t’ {aka ‘long long int’} [-Werror=format=]
 3037 |         snprintf(time, sizeof(time), "%s.%06ld", temp, now.tv_usec);
      |                                          ~~~~^         ~~~~~~~~~~~
      |                                              |            |
      |                                              long int     __suseconds64_t {aka long long int}
      |                                          %06lld
```

--- a/Makefile
+++ b/Makefile
@@ -28,6 +28,7 @@ CONDITIONAL_DEFS := \
  $(if $(value GLIBC_COMPAT),-DSFPTPD_GLIBC_COMPAT) \
  $(shell $(call if_header_then,sys/capability.h,-DHAVE_CAPS)) \
  $(shell $(call if_header_then,linux/ethtool_netlink.h,-DHAVE_ETHTOOL_NETLINK)) \
+ $(shell $(call if_header_then,linux/time_types.h,-DHAVE_TIME_TYPES)) \
  $(shell $(call if_defn_then,linux/if_link.h,IFLA_PERM_ADDRESS)) \
  $(shell $(call if_defn_then,linux/if_link.h,IFLA_PARENT_DEV_NAME))
 CONDITIONAL_LIBS := \

--- a/src/include/sfptpd_time.h
+++ b/src/include/sfptpd_time.h
@@ -4,7 +4,10 @@
 #ifndef _SFPTPD_TIME_H
 #define _SFPTPD_TIME_H
 
-#include <time.h>
+#ifdef HAVE_TIME_TYPES
+#include <linux/time_types.h>
+#endif
+#include <sys/time.h>
 #include <stdint.h>
 #include <stdbool.h>
 #include <math.h>
@@ -302,6 +305,19 @@ static inline void sfptpd_time_from_std_floor(struct sfptpd_timespec *sfts, cons
 	sfts->nsec = ts->tv_nsec;
 	sfts->nsec_frac = 0;
 }
+
+#ifdef HAVE_TIME_TYPES
+/** Convert a kernel timestamp to struct sfptpd_timespec.
+ * @param sfts  The structure to fill.
+ * @param ts The structure to copy.
+ */
+static inline void sfptpd_time_from_kernel_floor(struct sfptpd_timespec *sfts, const struct __kernel_timespec *ts)
+{
+	sfts->sec = ts->tv_sec;
+	sfts->nsec = ts->tv_nsec;
+	sfts->nsec_frac = 0;
+}
+#endif
 
 /** Initialise a struct sfptpd_timespec to zero.
  * @param ts  The structure to clear.

--- a/src/ptp/ptpd2/dep/msg.c
+++ b/src/ptp/ptpd2/dep/msg.c
@@ -3028,13 +3028,11 @@ appendSlaveStatusTLV(SlaveStatus *data, Octet *buf, size_t space)
 void msgDump(PtpInterface *ptpInterface)
 {
 	char temp[MAXTIMESTR], time[MAXTIMESTR + 8];
-	struct timeval now;
-	sfptpd_secs_t s;
+	struct sfptpd_timespec now;
 
-	gettimeofday(&now, 0);
-	s = (sfptpd_secs_t) now.tv_sec;
-	sfptpd_local_strftime(temp, sizeof(temp), "%Y-%m-%d %X", &s);
-	snprintf(time, sizeof(time), "%s.%06ld", temp, now.tv_usec);
+	sfclock_gettime(CLOCK_REALTIME, &now);
+	sfptpd_local_strftime(temp, sizeof(temp), "%Y-%m-%d %X", &now.sec);
+	snprintf(time, sizeof(time), "%s.%06" PRId32, temp, now.nsec / 1000);
 	time[sizeof(time) - 1] = '\0';
 
 	msgDebugHeader(&ptpInterface->msgTmpHeader, time);

--- a/src/ptp/ptpd2/dep/net.c
+++ b/src/ptp/ptpd2/dep/net.c
@@ -70,6 +70,8 @@
 #include <onload/extensions.h>
 #endif
 
+#include "sfptpd_time.h"
+
 /* SO_EE_ORIGIN_TIMESTAMPING is defined in linux/errqueue.h in recent kernels.
  * Define it for compilation with older kernels.
  */
@@ -90,8 +92,22 @@
 /* SO_TIMESTAMPING is defined in asm/socket.h in recent kernels.
  * Define it for compilation with older kernels.
  */
-#ifndef SO_TIMESTAMPING
-#define SO_TIMESTAMPING  37
+#ifndef SO_TIMESTAMPING_OLD
+#define SO_TIMESTAMPING_OLD  37
+#endif
+#ifndef SO_TIMESTAMPING_NEW
+#ifdef HAVE_TIME_TYPES
+struct scm_timestamping64 {
+        struct __kernel_timespec ts[3];
+};
+#endif
+#define SO_TIMESTAMPING_NEW  65
+#endif
+#ifndef SCM_TIMESTAMPING_OLD
+#define SCM_TIMESTAMPING_OLD SO_TIMESTAMPING_OLD
+#endif
+#ifndef SCM_TIMESTAMPING_NEW
+#define SCM_TIMESTAMPING_NEW SO_TIMESTAMPING_NEW
 #endif
 
 /* SCM_TIMESTAMPING_PKTINFO was introduced in Linux 4.13.
@@ -955,12 +971,15 @@ static void reset_timestamp(struct sfptpd_ts_info *info)
 static int parse_timestamp(uint8_t *pdu, size_t pdu_len,
 			   enum ptpd_ts_fmt ts_fmt,
 			   struct sfptpd_ts_info *info,
-			   bool tx)
+			   bool tx, int scm_type)
 {
 	const char *type = ts_fmt == PTPD_TS_ONLOAD_EXT ? "onload extension timestamp" : "so_timestamping";
 	union {
 		void *ptr;
-		struct timespec *lnx;
+		struct scm_timestamping *lnx;
+#ifdef HAVE_TIME_TYPES
+		struct scm_timestamping64 *lnx64;
+#endif
 #ifdef HAVE_ONLOAD_EXT
 		struct onload_timestamp *onload;
 #endif
@@ -972,7 +991,7 @@ static int parse_timestamp(uint8_t *pdu, size_t pdu_len,
 #ifdef HAVE_ONLOAD_EXT
 	    (ts_fmt == PTPD_TS_ONLOAD_EXT && pdu_len < CMSG_LEN(sizeof(*ts.onload) * 1)) ||
 #endif
-	    (ts_fmt == PTPD_TS_LINUX && pdu_len < CMSG_LEN(sizeof(*ts.lnx) * 3))) {
+	    (ts_fmt == PTPD_TS_LINUX && pdu_len < CMSG_LEN(sizeof(*ts.lnx)))) {
 		ERROR("received short %s (%zu)\n", type, pdu_len);
 		memset(info, '\0', sizeof *info);
 		return ENOTIMESTAMP;
@@ -994,9 +1013,16 @@ static int parse_timestamp(uint8_t *pdu, size_t pdu_len,
 		sfptpd_time_zero(&info->sw);
 		info->have_sw = false;
 #endif
+#ifdef HAVE_TIME_TYPES
+	} else if (scm_type == SCM_TIMESTAMPING_NEW) {
+		sfptpd_time_from_kernel_floor(&info->sw, &ts.lnx64->ts[0]);
+		sfptpd_time_from_kernel_floor(&info->hw, &ts.lnx64->ts[2]);
+		info->have_sw = info->sw.sec != 0;
+		info->have_hw = info->hw.sec != 0;
+#endif
 	} else {
-		sfptpd_time_from_std_floor(&info->sw, &ts.lnx[0]);
-		sfptpd_time_from_std_floor(&info->hw, &ts.lnx[2]);
+		sfptpd_time_from_std_floor(&info->sw, &ts.lnx->ts[0]);
+		sfptpd_time_from_std_floor(&info->hw, &ts.lnx->ts[2]);
 		info->have_sw = info->sw.sec != 0;
 		info->have_hw = info->hw.sec != 0;
 	}
@@ -1121,10 +1147,11 @@ bool netProcessError(PtpInterface *ptpInterface,
 		DUMP("cmsg (header)", cmsg, len);
 
 		if ((SOL_SOCKET == level) &&
-		    (SO_TIMESTAMPING == type)) {
+		    (SO_TIMESTAMPING_OLD == type ||
+		     SO_TIMESTAMPING_NEW == type)) {
 			haveTs = parse_timestamp(CMSG_DATA(cmsg), len,
 						 ptpInterface->ts_fmt,
-						 info, true) == 0;
+						 info, true, type) == 0;
 		} else if ((ipproto == level) &&
 			   (iptype == type)) {
 			err = (struct sock_extended_err *)CMSG_DATA(cmsg);
@@ -1369,10 +1396,11 @@ static Boolean getRxTimestamp(PtpInterface *ptpInterface,
 			info->have_sw = true;
 			break;
 
-		case SO_TIMESTAMPING:
+		case SCM_TIMESTAMPING_NEW:
+		case SCM_TIMESTAMPING_OLD:
 			parse_timestamp(CMSG_DATA(cmsg), cmsg->cmsg_len,
 					ptpInterface->ts_fmt,
-					info, FALSE);
+					info, FALSE, cmsg->cmsg_type);
 			break;
 
 		case SCM_TIMESTAMPING_PKTINFO:
@@ -1397,7 +1425,8 @@ static Boolean getRxTimestamp(PtpInterface *ptpInterface,
 
 /* Attempt to use SO_TIMESTAMPING with the provided flags, and try to include
  * SOF_TIMESTAMPING_OPT_PKTINFO if possible. */
-static Boolean netTryEnableTimestampingPktinfo(PtpInterface *ptpInterface, int flags)
+static Boolean netTryEnableTimestampingPktinfo(PtpInterface *ptpInterface,
+					       int flags, int type)
 {
 	int has_pktinfo_flag;
 
@@ -1418,6 +1447,22 @@ static Boolean netTryEnableTimestampingPktinfo(PtpInterface *ptpInterface, int f
 	return FALSE;
 }
 
+/* y2038: attempt SO_TIMESTAMPING_NEW, falling back to SO_TIMESTAMPING_OLD,
+ * allowing 32-bit systems to access 64-bit seconds on recent kernels. */
+static bool netTryEnableTimestamping(PtpInterface *ptpInterface, int flags)
+{
+	const int option[] = {
+		SO_TIMESTAMPING_NEW,
+		SO_TIMESTAMPING_OLD,
+	};
+	enum { FIRST, LAST } iter;
+	bool done = false;
+
+	for (iter = FIRST; !done && iter <= LAST; iter++)
+		done = netTryEnableTimestampingPktinfo(ptpInterface, flags, option[iter]);
+
+	return done;
+}
 
 /**
  * Initialize timestamping of packets
@@ -1461,7 +1506,7 @@ netInitTimestamping(PtpInterface *ptpInterface, InterfaceOpts *ifOpts)
 		      | SOF_TIMESTAMPING_RX_SOFTWARE
 		      | SOF_TIMESTAMPING_SOFTWARE;
 
-		if (netTryEnableTimestampingPktinfo(ptpInterface, flags)) {
+		if (netTryEnableTimestamping(ptpInterface, flags)) {
 			ptpInterface->tsMethod = TS_METHOD_SO_TIMESTAMPING;
 			INFO("using SO_TIMESTAMPING software timestamps\n");
 			return TRUE;
@@ -1538,7 +1583,7 @@ netInitTimestamping(PtpInterface *ptpInterface, InterfaceOpts *ifOpts)
 	      | SOF_TIMESTAMPING_RX_HARDWARE
 	      | SOF_TIMESTAMPING_RAW_HARDWARE;
 
-	if (netTryEnableTimestampingPktinfo(ptpInterface, flags)) {
+	if (netTryEnableTimestamping(ptpInterface, flags)) {
 		ptpInterface->tsMethod = TS_METHOD_SO_TIMESTAMPING;
 		INFO("using SO_TIMESTAMPING hardware timestamps\n");
 		return TRUE;

--- a/src/ptp/ptpd2/dep/net.c
+++ b/src/ptp/ptpd2/dep/net.c
@@ -79,19 +79,34 @@
 #define SO_EE_ORIGIN_TIMESTAMPING 4
 #endif
 
-/* SO_TIMESTAMP is defined in asm/socket.h in recent kernels.
- * Define it here for compilation with older kernels.
- */
-#ifndef SO_TIMESTAMPNS
-#define SO_TIMESTAMPNS 35
-#ifndef SCM_TIMESTAMPNS
-#define SCM_TIMESTAMPNS SO_TIMESTAMPNS
+/* The _NEW and _OLD versions of SO_TIMESTAMP* options are only defined in
+ * quite recent kernels. Define them here for compatibility.
+ *
+ * The _NEW variants are for 64-bit timestamps on 32-bit systems. The _OLD
+ * variant uses the architecture-native type. */
+
+#define SCM(option) SO_ ## option
+
+/* The SO_TIMESTAMP* microsecond sw timestamping options */
+
+#ifndef SO_TIMESTAMP_OLD
+#define SO_TIMESTAMP_OLD 29
 #endif
+#ifndef SO_TIMESTAMP_NEW
+#define SO_TIMESTAMP_NEW 63
 #endif
 
-/* SO_TIMESTAMPING is defined in asm/socket.h in recent kernels.
- * Define it for compilation with older kernels.
- */
+/* The SO_TIMESTAMPNS* nanosecond sw timestamping options */
+
+#ifndef SO_TIMESTAMPNS_OLD
+#define SO_TIMESTAMPNS_OLD 35
+#endif
+#ifndef SO_TIMESTAMPNS_NEW
+#define SO_TIMESTAMPNS_NEW 64
+#endif
+
+/* The SO_TIMESTAMPING* nanosecond hw/sw timestamping options */
+
 #ifndef SO_TIMESTAMPING_OLD
 #define SO_TIMESTAMPING_OLD  37
 #endif
@@ -102,12 +117,6 @@ struct scm_timestamping64 {
 };
 #endif
 #define SO_TIMESTAMPING_NEW  65
-#endif
-#ifndef SCM_TIMESTAMPING_OLD
-#define SCM_TIMESTAMPING_OLD SO_TIMESTAMPING_OLD
-#endif
-#ifndef SCM_TIMESTAMPING_NEW
-#define SCM_TIMESTAMPING_NEW SO_TIMESTAMPING_NEW
 #endif
 
 /* SCM_TIMESTAMPING_PKTINFO was introduced in Linux 4.13.
@@ -135,6 +144,76 @@ struct scm_ts_pktinfo {
 #include <net-snmp/net-snmp-includes.h>
 #include <net-snmp/agent/net-snmp-agent-includes.h>
 #endif
+
+union scm_ts_payload_ptr {
+	void *ptr;
+	struct timeval *usec;
+	struct timespec *singleton;
+	struct scm_timestamping *triplet;
+#ifdef HAVE_TIME_TYPES
+	struct __kernel_sock_timeval *usec64;
+	struct __kernel_timespec *singleton64;
+	struct scm_timestamping64 *triplet64;
+#endif
+#ifdef HAVE_ONLOAD_EXT
+	struct onload_timestamp *onload;
+#endif
+};
+
+static const char *get_timestamp_name(enum ptpd_ts_fmt ts_fmt,
+				      int so_option)
+{
+	if (ts_fmt == PTPD_TS_ONLOAD_EXT) {
+		return "onload extension";
+	}
+	switch (so_option) {
+	case SO_TIMESTAMP_OLD:
+		return "SO_TIMESTAMP_OLD";
+	case SO_TIMESTAMP_NEW:
+		return "SO_TIMESTAMP_NEW";
+	case SO_TIMESTAMPNS_OLD:
+		return "SO_TIMESTAMPNS_OLD";
+	case SO_TIMESTAMPNS_NEW:
+		return "SO_TIMESTAMPNS_NEW";
+	case SO_TIMESTAMPING_OLD:
+		return "SO_TIMESTAMPING_OLD";
+	case SO_TIMESTAMPING_NEW:
+		return "SO_TIMESTAMPING_NEW";
+	default:
+		return "SO_?";
+	}
+}
+
+size_t get_timestamp_size(enum ptpd_ts_fmt ts_fmt,
+			  int scm_type)
+{
+	union scm_ts_payload_ptr ts;
+	if (ts_fmt == PTPD_TS_ONLOAD_EXT) {
+#ifdef HAVE_ONLOAD_EXT
+		return sizeof *ts.onload;
+#else
+		return SIZE_MAX;
+#endif
+	}
+	switch (scm_type) {
+	case SCM(TIMESTAMPING_OLD):
+		return sizeof *ts.triplet;
+	case SCM(TIMESTAMPNS_OLD):
+		return sizeof *ts.singleton;
+	case SCM(TIMESTAMP_OLD):
+		return sizeof *ts.usec;
+#ifdef HAVE_TIME_TYPES
+	case SCM(TIMESTAMPNS_NEW):
+		return sizeof *ts.singleton64;
+	case SCM(TIMESTAMPING_NEW):
+		return sizeof *ts.triplet64;
+	case SCM(TIMESTAMP_NEW):
+		return sizeof *ts.usec64;
+#endif
+	default:
+		return SIZE_MAX;
+	}
+}
 
 
 /****************************************************************************
@@ -973,31 +1052,21 @@ static int parse_timestamp(uint8_t *pdu, size_t pdu_len,
 			   struct sfptpd_ts_info *info,
 			   bool tx, int scm_type)
 {
-	const char *type = ts_fmt == PTPD_TS_ONLOAD_EXT ? "onload extension timestamp" : "so_timestamping";
-	union {
-		void *ptr;
-		struct scm_timestamping *lnx;
-#ifdef HAVE_TIME_TYPES
-		struct scm_timestamping64 *lnx64;
-#endif
-#ifdef HAVE_ONLOAD_EXT
-		struct onload_timestamp *onload;
-#endif
-	} ts;
+	union scm_ts_payload_ptr ts;
+	const char *type;
 
 	assert(info);
 
-	if (
-#ifdef HAVE_ONLOAD_EXT
-	    (ts_fmt == PTPD_TS_ONLOAD_EXT && pdu_len < CMSG_LEN(sizeof(*ts.onload) * 1)) ||
-#endif
-	    (ts_fmt == PTPD_TS_LINUX && pdu_len < CMSG_LEN(sizeof(*ts.lnx)))) {
+	type = get_timestamp_name(ts_fmt, scm_type);
+	if (pdu_len < CMSG_LEN(get_timestamp_size(ts_fmt, scm_type))) {
 		ERROR("received short %s (%zu)\n", type, pdu_len);
 		memset(info, '\0', sizeof *info);
 		return ENOTIMESTAMP;
 	}
-
 	ts.ptr = pdu;
+
+	sfptpd_time_zero(&info->sw);
+	sfptpd_time_zero(&info->hw);
 
 	if (ts_fmt == PTPD_TS_ONLOAD_EXT) {
 #ifdef HAVE_ONLOAD_EXT
@@ -1005,27 +1074,28 @@ static int parse_timestamp(uint8_t *pdu, size_t pdu_len,
 			info->hw.sec = ts.onload->sec;
 			info->hw.nsec = ts.onload->nsec;
 			info->hw.nsec_frac = (uint32_t) ts.onload->nsec_frac << 8;
-			info->have_hw = true;
-		} else {
-			sfptpd_time_zero(&info->hw);
-			info->have_hw = false;
 		}
-		sfptpd_time_zero(&info->sw);
-		info->have_sw = false;
 #endif
 #ifdef HAVE_TIME_TYPES
-	} else if (scm_type == SCM_TIMESTAMPING_NEW) {
-		sfptpd_time_from_kernel_floor(&info->sw, &ts.lnx64->ts[0]);
-		sfptpd_time_from_kernel_floor(&info->hw, &ts.lnx64->ts[2]);
-		info->have_sw = info->sw.sec != 0;
-		info->have_hw = info->hw.sec != 0;
+	} else if (scm_type == SCM(TIMESTAMPING_NEW)) {
+		sfptpd_time_from_kernel_floor(&info->sw, &ts.triplet64->ts[0]);
+		sfptpd_time_from_kernel_floor(&info->hw, &ts.triplet64->ts[2]);
+	} else if (scm_type == SCM(TIMESTAMPNS_NEW)) {
+		sfptpd_time_from_kernel_floor(&info->sw, ts.singleton64);
+	} else if (scm_type == SCM(TIMESTAMP_NEW)) {
+		sfptpd_time_init(&info->sw, ts.usec64->tv_sec, ts.usec64->tv_usec * 1000, 0);
 #endif
-	} else {
-		sfptpd_time_from_std_floor(&info->sw, &ts.lnx->ts[0]);
-		sfptpd_time_from_std_floor(&info->hw, &ts.lnx->ts[2]);
-		info->have_sw = info->sw.sec != 0;
-		info->have_hw = info->hw.sec != 0;
+	} else if (scm_type == SCM(TIMESTAMPING_OLD)) {
+		sfptpd_time_from_std_floor(&info->sw, &ts.triplet->ts[0]);
+		sfptpd_time_from_std_floor(&info->hw, &ts.triplet->ts[2]);
+	} else if (scm_type == SCM(TIMESTAMPNS_OLD)) {
+		sfptpd_time_from_std_floor(&info->sw, ts.singleton);
+	} else if (scm_type == SCM(TIMESTAMP_OLD)) {
+		sfptpd_time_init(&info->sw, ts.usec->tv_sec, ts.usec->tv_usec * 1000, 0);
 	}
+
+	info->have_sw = info->sw.sec != 0;
+	info->have_hw = info->hw.sec != 0;
 
 	if (info->have_sw) {
 		DBG("%s sw timestamp: " SFPTPD_FMT_SFTIMESPEC " (%s)\n",
@@ -1362,8 +1432,6 @@ static Boolean getRxTimestamp(PtpInterface *ptpInterface,
 	}
 
 	for (cmsg = CMSG_FIRSTHDR(msg); cmsg != NULL; cmsg = CMSG_NXTHDR(msg, cmsg)) {
-		struct timeval *tv;
-		struct timespec *ts;
 		struct scm_ts_pktinfo *ts_pktinfo;
 
 		DUMP("CM", cmsg, cmsg->cmsg_len);
@@ -1372,32 +1440,12 @@ static Boolean getRxTimestamp(PtpInterface *ptpInterface,
 			continue;
 
 		switch (cmsg->cmsg_type) {
-		case SCM_TIMESTAMP:
-			tv = (struct timeval *)CMSG_DATA(cmsg);
-
-			if(cmsg->cmsg_len < CMSG_LEN(sizeof(*tv))) {
-				ERROR("received short SCM_TIMESTAMP (%zu)\n",
-				      cmsg->cmsg_len);
-				return FALSE;
-			}
-			sfptpd_time_init(&info->sw, tv->tv_sec, tv->tv_usec * 1000, 0);
-			info->have_sw = true;
-			break;
-
-		case SCM_TIMESTAMPNS:
-			ts = (struct timespec *)CMSG_DATA(cmsg);
-
-			if(cmsg->cmsg_len < CMSG_LEN(sizeof(*ts))) {
-				ERROR("received short SCM_TIMESTAMPNS (%zu)\n",
-				      cmsg->cmsg_len);
-				return FALSE;
-			}
-			sfptpd_time_init(&info->sw, ts->tv_sec, ts->tv_nsec, 0);
-			info->have_sw = true;
-			break;
-
-		case SCM_TIMESTAMPING_NEW:
-		case SCM_TIMESTAMPING_OLD:
+		case SCM(TIMESTAMP_NEW):
+		case SCM(TIMESTAMP_OLD):
+		case SCM(TIMESTAMPNS_NEW):
+		case SCM(TIMESTAMPNS_OLD):
+		case SCM(TIMESTAMPING_NEW):
+		case SCM(TIMESTAMPING_OLD):
 			parse_timestamp(CMSG_DATA(cmsg), cmsg->cmsg_len,
 					ptpInterface->ts_fmt,
 					info, FALSE, cmsg->cmsg_type);
@@ -1464,6 +1512,34 @@ static bool netTryEnableTimestamping(PtpInterface *ptpInterface, int flags)
 	return done;
 }
 
+/* y2038: attempt SO_TIMESTAMP* family of options in pref order. */
+static bool netTryEnableTimestamp(PtpInterface *ptpInterface)
+{
+	const int prefs[] = {
+		SO_TIMESTAMPNS_NEW,
+		SO_TIMESTAMPNS_OLD,
+		SO_TIMESTAMP_NEW,
+		SO_TIMESTAMP_OLD,
+	};
+	const int num_options = sizeof prefs / sizeof prefs[0];
+	int iter;
+
+	for (iter = 0; iter < num_options; iter++) {
+		int option = prefs[iter];
+		int flags = 1;
+		DBG("trying %s software timestamping...\n",
+		    get_timestamp_name(PTPD_TS_LINUX, option));
+		if (setsockopt(ptpInterface->transport.eventSock, SOL_SOCKET,
+			       option, &flags, sizeof(flags)) == 0) {
+			ptpInterface->tsMethod = TS_METHOD_SYSTEM;
+			INFO("using %s software timestamps\n",
+			     get_timestamp_name(PTPD_TS_LINUX, option));
+			break;
+		}
+	}
+	return iter != num_options;
+}
+
 /**
  * Initialize timestamping of packets
  *
@@ -1514,30 +1590,8 @@ netInitTimestamping(PtpInterface *ptpInterface, InterfaceOpts *ifOpts)
 	}
 
 	if (ifOpts->timestampType == PTPD_TIMESTAMP_TYPE_SW) {
-
-		/* Try SO_TIMESTAMPNS software timestamps */
-		DBG("trying SO_TIMESTAMPNS software timestamping...\n");
-
-		/* Enable software timestamps */
-		flags = 1;
-
-		if (setsockopt(ptpInterface->transport.eventSock, SOL_SOCKET,
-			       SO_TIMESTAMPNS, &flags, sizeof(flags)) == 0) {
+		if (netTryEnableTimestamp(ptpInterface)) {
 			ptpInterface->tsMethod = TS_METHOD_SYSTEM;
-			INFO("using SO_TIMESTAMPNS software timestamps\n");
-			return TRUE;
-		}
-
-		/* Try SO_TIMESTAMP software timestamps */
-		DBG("trying SO_TIMESTAMP software timestamping...\n");
-
-		/* Enable software timestamps */
-		flags = 1;
-
-		if (setsockopt(ptpInterface->transport.eventSock, SOL_SOCKET,
-			       SO_TIMESTAMP, &flags, sizeof(flags)) == 0) {
-			ptpInterface->tsMethod = TS_METHOD_SYSTEM;
-			INFO("using SO_TIMESTAMP software timestamps\n");
 			return TRUE;
 		}
 

--- a/src/sfptpd_logging.c
+++ b/src/sfptpd_logging.c
@@ -1066,18 +1066,17 @@ struct sfptpd_log *sfptpd_log_open_remote_monitor(void)
 
 void sfptpd_log_get_time(struct sfptpd_log_time *time)
 {
-	struct timeval now;
 	char temp[SFPTPD_LOG_TIME_STR_MAX];
-	sfptpd_secs_t s;
+	struct sfptpd_timespec now;
 	int rc;
 	
 	assert(time != NULL);
 	
-	gettimeofday(&now, 0);
-	s = (sfptpd_secs_t) now.tv_sec;
-	sfptpd_local_strftime(temp, sizeof(temp), "%Y-%m-%d %X", &s);
+	sfclock_gettime(CLOCK_REALTIME, &now);
+	sfptpd_local_strftime(temp, sizeof(temp), "%Y-%m-%d %X", &now.sec);
 
-	rc = snprintf(time->time, sizeof(time->time), "%s.%06ld", temp, now.tv_usec);
+	rc = snprintf(time->time, sizeof(time->time), "%s.%06" PRId32,
+		      temp, now.nsec / 1000);
 	assert(rc < sizeof(time->time));
 }
 

--- a/src/sfptpd_main.c
+++ b/src/sfptpd_main.c
@@ -252,9 +252,10 @@ static int drop_user(struct sfptpd_config *config)
 
 static int runtime_checks(struct sfptpd_config *config)
 {
-	FILE *clock_src;
+	struct sfptpd_config_general *gconf = sfptpd_general_config_get(config);
 	char source[10] = "";
 	struct utsname name;
+	FILE *clock_src;
 	int rc = 0;
 
 	assert(config != NULL);
@@ -262,13 +263,11 @@ static int runtime_checks(struct sfptpd_config *config)
 	rc = uname(&name);
 #if defined(__i386__)
 	/* If this is a 32bit binary, check that we are running on a 32bit kernel */
-	if (rc == -1) {
-		CRITICAL("could not determine system characteristics with uname: %s\n",
-			 strerror(errno));
-		return ENOEXEC;
-	} else if (strcmp(name.machine, "i686") != 0) {
-		CRITICAL("32-bit sfptpd not compatible with 64-bit kernel\n");
-		return ENOEXEC;
+	if (rc == 0 &&
+	    strcmp(name.machine, "i686") != 0 &&
+	    !gconf->avoid_efx) {
+		gconf->avoid_efx = true;
+		NOTICE("disabling private sfc clock ioctls with 32-bit userspace on 64-bit kernel\n");
 	}
 #endif
 
@@ -284,7 +283,7 @@ static int runtime_checks(struct sfptpd_config *config)
 #endif
 	}
 
-	if (sfptpd_general_config_get(config)->lock) {
+	if (gconf->lock) {
 		struct sfptpd_prog competitors[] = {
 			{ "ptpd*" },
 			{ "sfptpd" },

--- a/src/sfptpd_phc.c
+++ b/src/sfptpd_phc.c
@@ -33,18 +33,24 @@
 #include "sfptpd_priv.h"
 
 
+#ifdef SFPTPD_GLIBC_COMPAT
 /****************************************************************************
  * Missing kernel API bits and pieces
  ****************************************************************************/
 
 /* The oldest distribution we support has kernel support for PHC but the
  * syscall is missing in that glibc version. In this case we need to create our
- * own version of clock_adjtime() and invoke the syscall. */
+ * own version of clock_adjtime() and invoke the syscall.
+ *
+ * Avoid including this where not needed as it overrides glibc's smarts for
+ * handling 64-bit time on 32-bit architectures.
+ */
 #pragma weak clock_adjtime
 int clock_adjtime(clockid_t clock, struct timex *timex_block)
 {
 	return syscall(__NR_clock_adjtime, clock, timex_block);
 }
+#endif
 
 
 /****************************************************************************

